pkg/stanza/fileconsumer/file.go

```diff
@@ -182,82 +192,67 @@ func (m *Manager) consume(ctx context.Context, paths []string) {
-	files := make([]*os.File, 0, len(filesPaths))
-	for _, path := range filesPaths {
-		if _, ok := m.seenPaths[path]; !ok {
-			if m.readerFactory.fromBeginning {
-				m.Infow("Started watching file", "path", path)
-			} else {
-				m.Infow("Started watching file from end. To read preexisting logs, configure the argument 'start_at' to 'beginning'", "path", path)
-			}
-			m.seenPaths[path] = struct{}{}
-		}
-		file, err := os.Open(path) // #nosec - operator must read in files defined by user
-		if err != nil {
-			m.Debugf("Failed to open file", zap.Error(err))
-			continue
+	if _, ok := m.seenPaths[path]; !ok {
+		if m.readerFactory.fromBeginning {
+			m.Infow("Started watching file", "path", path)
+		} else {
+			m.Infow("Started watching file from end. To read preexisting logs, configure the argument 'start_at' to 'beginning'", "path", path)
 		}
-		files = append(files, file)
+		m.seenPaths[path] = struct{}{}
+	}
+	file, err := os.Open(path) // #nosec - operator must read in files defined by user
+	if err != nil {
+		m.Errorf("Failed to open file", zap.Error(err))
+		return nil
 	}
 
 	// Get fingerprints for each file
-	fps := make([]*Fingerprint, 0, len(files))
-	for _, file := range files {
-		fp, err := m.readerFactory.newFingerprint(file)
-		if err != nil {
-			m.Errorw("Failed creating fingerprint", zap.Error(err))
-			continue
+	fp, err := m.readerFactory.newFingerprint(file)
+	if err != nil {
+		m.Errorw("Failed creating fingerprint", zap.Error(err))
+		return nil
+	}
+
+	if len(fp.FirstBytes) == 0 {
+		// Empty file, don't read it until we can compare its fingerprint
+		if err = file.Close(); err != nil {
+			m.Errorf("problem closing file %s", file.Name())
+			return nil
 		}
-		fps = append(fps, fp)
 	}
+	m.currentFps = append(m.currentFps, fp)
 
 	// Exclude any empty fingerprints or duplicate fingerprints to avoid doubling up on copy-truncate files
-OUTER:
-	for i := 0; i < len(fps); i++ {
-		fp := fps[i]
-		if len(fp.FirstBytes) == 0 {
-			if err := files[i].Close(); err != nil {
-				m.Errorf("problem closing file %s", files[i].Name())
-			}
-			// Empty file, don't read it until we can compare its fingerprint
-			fps = append(fps[:i], fps[i+1:]...)
-			files = append(files[:i], files[i+1:]...)
-			i--
-			continue
-		}
-		for j := i + 1; j < len(fps); j++ {
-			fp2 := fps[j]
-			if fp.StartsWith(fp2) || fp2.StartsWith(fp) {
-				// Exclude
-				if err := files[i].Close(); err != nil {
-					m.Errorf("problem closing file %s", files[i].Name())
-				}
-				fps = append(fps[:i], fps[i+1:]...)
-				files = append(files[:i], files[i+1:]...)
-				i--
-				continue OUTER
+	for i := 0; i < len(m.currentFps)-1; i++ {
+		fp2 := m.currentFps[i]
+		if fp.StartsWith(fp2) || fp2.StartsWith(fp) {
+			// Exclude duplicates
+			if err = file.Close(); err != nil {
+				m.Errorf("problem closing file", "file", file.Name())
 			}
+			m.currentFps = m.currentFps[:len(m.currentFps)-1]
```

Comment:
We're adding this just above, then having to adjust the iteration bounds to account for it, then possibly removing it. Why not just add it after we've determined we actually want to keep it?

Reply:
Addressed this change